### PR TITLE
feat: Add decimal setting kind

### LINKED ISSUE #3424: Consider adding a new setting kind `decimal`
Migrated from GitLab: https://gitlab.com/meltano/meltano/-/issues/3508

Originally created by @edgarrmondragon on 2022-05-19 01:16:39

---

### Problem to solve

Allow developers to declare plugins with numeric settings that don't really fit any of the currently supported types.

### Target audience

Plugin developers that require numeric settings.

### Further details

One example that comes to mind is Unix timestamps with decimal precision.

### Proposal

Meltano currently supports the following setting types:

- oauth
- hidden
- password
- date_iso8601
- file
- email
- integer
- options
- object
- array
- boolean
- string

A `decimal` type, cast to Python's `decimal.Decimal` is easy to add. Alternatives to the `decimal` name are: `real`, `number`, `numeric`.

### What does success look like, and how can we measure that?

A new setting kind with support for 

### Regression test

- [ ] Write adequate test cases and submit test results
- [ ] Test results should be reviewed by a person from the team

### Links / references

- Cast logic: https://gitlab.com/meltano/meltano/-/blob/6d96c9bc9ec5610d87dc01d568bbdb3b162210a4/src/meltano/core/setting_definition.py#L188-212
- `SettingKind` enum: https://gitlab.com/meltano/meltano/-/blob/6d96c9bc9ec5610d87dc01d568bbdb3b162210a4/src/meltano/core/setting_definition.py#L62-76

### PR DESCRIPTION
**Pull request by mahangu**
## Description

  Adds a new `decimal` setting kind to support numeric values with decimal precision. This enables plugin developers to handle values like Unix timestamps with fractional seconds
  (e.g., `1640995200.123`) that don't fit the existing `integer` or `string` types.

  **Implementation:**
  - Adds `DECIMAL` to the `SettingKind` enum
  - Implements casting logic using Python's `decimal.Decimal` for precise arithmetic
  - Adds automatic type inference for both `Decimal` and `float` values
  - Updates JSON schema validation to accept `decimal` as a valid setting kind
  - Includes comprehensive test coverage for edge cases like scientific notation and precision handling

  The new setting type follows the same patterns as existing types and integrates seamlessly with Meltano's configuration system.

## Help!

This is my first Meltano PR so please let me know if I've missed anything. 🙏🏾 Thank you!

  ## Related Issues

  * Closes #3424

## Summary by Sourcery

Add support for a new decimal setting kind to handle numeric values with fractional precision in Meltano configurations.

New Features:
- Introduce DECIMAL to SettingKind with casting logic using Python's decimal.Decimal
- Enable automatic kind inference from Decimal and float values
- Update JSON schema validation to accept decimal as a valid setting kind
- Extend stringify_value to handle decimal settings as strings

Tests:
- Add tests for casting decimal strings (including scientific notation and invalid formats)
- Add tests for non-string decimal values passing through unchanged
- Add tests for automatic decimal type inference from Decimal and float inputs
- Add tests for correct stringification and round-trip casting of decimal values

### CODE REVIEW

File: tests/meltano/core/test_setting_definition.py [hunk 1]
```diff
@@ -75,6 +76,85 @@ def test_cast_options(self) -> None:
...
+
+    def test_cast_value_decimal_non_string(self) -> None:
+        """Test that non-string values are passed through for decimal settings."""
+        setting_definition = SettingDefinition(
+            name="test_setting", kind=SettingKind.DECIMAL
+        )
+
+        decimal_val = Decimal("3.14")
+        assert setting_definition.cast_value(decimal_val) == decimal_val
+        assert setting_definition.cast_value(None) is None
```
[the remark this answers is not in the document]
Comment (mahangu) [reply tests/meltano/core/test_setting_definition.py:121]: Thanks for the suggestion! However from what I can see, this suggested test step fails atm because:

  1. **`cast_value()` seems to be specifically for string-to-type conversion** (from env vars, YAML, etc.), not general type-to-type conversion. Here's what happens with the same float value:
     - `INTEGER.cast_value("3.14")` → `3` (int from string)
     - `DECIMAL.cast_value("3.14")` → `Decimal("3.14")` (Decimal from string)

     Non-strings are returned unchanged regardless of setting kind:
     - `INTEGER.cast_value(3.14)` → `3.14` (float unchanged)
     - `DECIMAL.cast_value(3.14)` → `3.14` (float unchanged)

  2. **Float → Decimal mapping happens in `from_key_value()`** for type inference, which we already test in `test_from_key_value_decimal_inference()`:
     ```python
     float_setting = SettingDefinition.from_key_value("float_key", 3.14)
     assert float_setting.kind == SettingKind.DECIMAL  # ← This tests float inference

I could be missing something here though. Let me know what you think!

**Comment by tayloramurphy** [comment]
@mahangu we'll need a docs update to https://docs.meltano.com/reference/plugin-definition-syntax#settingskind at least as well. I'm not 100% if these options are mentioned elsewhere.

**Comment by mahangu** [comment]
Noted, @tayloramurphy! Thank you. Found one other instance in `docs/` and updated both in [bc74603](https://github.com/meltano/meltano/pull/9374/commits/bc746034d225bad44f08892cb0b9958da420f0db).

**Review by edgarrmondragon — APPROVED** [review]
Thanks @mahangu!

**Comment by edgarrmondragon** [comment]
@all-contributors please add @mahangu for code
